pkg/storegateway/indexheader/index/postings.go

```diff
@@ -87,58 +87,105 @@ func newV1PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset
 	return &t, nil
 }
 
-func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (*PostingOffsetTableV2, error) {
+type lastValuePosition struct {
+	key                string
+	valueOffsetFromEnd int
+	offsetInTable      int
+}
+
+func newV2PostingOffsetTable(factory *streamencoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (table *PostingOffsetTableV2, err error) {
 	t := PostingOffsetTableV2{
 		factory:                     factory,
 		tableOffset:                 tableOffset,
 		postings:                    map[string]*postingValueOffsets{},
 		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
 	}
 
-	lastTableOff := 0
-	valueCount := 0
-	var lastKey string
-	var lastValue string
+	d := factory.NewDecbufAtChecked(tableOffset, castagnoliTable)
+	defer runutil.CloseWithErrCapture(&err, &d, "read offset table")
 
-	// For the postings offset table we keep every label name but only every nth
-	// label value (plus the first and last one), to save memory.
-	if err := readOffsetTable(factory, tableOffset, func(key string, value string, off uint64, tableOff int) error {
-		if _, ok := t.postings[key]; !ok {
-			// Not seen before label name.
-			if len(t.postings) > 0 {
-				// Always include last value for each label name, unless it was just added in previous iteration based
-				// on valueCount.
-				if (valueCount-1)%postingOffsetsInMemSampling != 0 {
-					t.postings[lastKey].offsets = append(t.postings[lastKey].offsets, postingOffset{value: lastValue, tableOff: lastTableOff})
-				}
+	startLen := d.Len()
+	remainingCount := d.Be32()
+	currentKey := ""
+	valuesForCurrentKey := 0
+
+	var lastValuePositions []lastValuePosition
+
+	for d.Err() == nil && remainingCount > 0 {
+		lastKey := currentKey
+		offsetInTable := startLen - d.Len()
+		keyCount := d.Uvarint()
+
+		// The Postings offset table takes only 2 keys per entry (name and value of label).
+		if keyCount != 2 {
+			return nil, errors.Errorf("unexpected key length for posting table %d", keyCount)
+		}
+
+		// Important: this value is only valid as long as we don't perform any further reads from d.
+		// If we need to retain its value, we must copy it before performing another read.
+		key := d.UvarintBytes()
+
+		if len(t.postings) == 0 || currentKey != string(key) {
+			// New label name.
+			currentKey = string(key)
+			t.postings[currentKey] = &postingValueOffsets{}
+			lastValuePositions = append(lastValuePositions, lastValuePosition{key: currentKey, valueOffsetFromEnd: -1})
+			valuesForCurrentKey = 0
+		}
+
+		if valuesForCurrentKey%postingOffsetsInMemSampling == 0 {
```

Comment:
[nit] I always get tricked by this math. Could you add a comment to explain this is always true for the 1st value, please?

Reply:
Resolved in 280ad1b.